base/compiler/abstractinterpretation.jl

```diff
@@ -879,10 +879,21 @@ function abstract_eval(@nospecialize(e), vtypes::VarTable, sv::InferenceState)
         argtypes = Any[ abstract_eval(a, vtypes, sv) for a in e.args ]
         t = abstract_eval_call(e.args, argtypes, vtypes, sv)
     elseif e.head === :new
-        t = instanceof_tfunc(abstract_eval(e.args[1], vtypes, sv))[1]
-        for i = 2:length(e.args)
-            if abstract_eval(e.args[i], vtypes, sv) === Bottom
-                rt = Bottom
+        argtypes = Any[ abstract_eval(a, vtypes, sv) for a in e.args ]
+        t = instanceof_tfunc(argtypes[1])[1]
+        isconst = isbitstype(t)
+        for i = 2:length(argtypes)
+            ae = argtypes[i]
+            if ae === Bottom
+                t = Bottom
+            end
+            isconst &= ae isa Const
+        end
+        if isconst
+            flds = Any[ argtypes[i].val for i = 2:length(argtypes) ]
+            try
+                t = Const(ccall(:jl_new_structv, Any, (Any, Ptr{Cvoid}, UInt32), t, flds, length(flds)))
+            catch ex
```

Comment:
It would be better to check for the error conditions explicitly; it's faster than try/catch and would allow returning `Bottom` for errors. I think all you need to check is that each value is of the corresponding field type. Elsewhere we assume that the number of arguments is correct.

Reply:
I believe that's correct. Plus it technically lets you do cool thinks like return a Const inferred value even if one of the values is only inferred as a union (as long as one of the union elements matches precisely - and they're both singletons at the moment).

Reply:
I was worried there might be another failure possibility (OOM? others?), but yeah, I guess it's just the type error, and we can check for that in the for-loop.

IIRC, the number of arguments is checked syntactically, so it would be invalid IR to discover otherwise.

Reply:
Okay, I'm still a little new at this and I couldn't figure it out after looking around: anyone know which function would give the field types of the struct?

Reply:
`fieldtype(type, index)` gives you the type of a field.

Reply:
Ah, yes, thanks! All fixed up, I think.